### PR TITLE
Make all parameter names in `@overload`s match the API being overloaded.

### LINKED ISSUE #9053: Numpy Gamma distribution not working in Numba 0.57+
## Reporting a bug

- [ x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/main/CHANGE_LOG).
- [ x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

Vectorization of Numpy's gamma distribution no longer works. The below code works in Numba 0.56.4, but not in 0.57.0 or 0.57.1. 

```
from numba import float64, vectorize
from numpy.random import gamma
import numpy as np

@vectorize([float64(float64, float64)])
def vec_ig(shape, scale):
    ig = 1. / gamma(shape=shape, scale=1. / scale)
    return ig

a = np.array([0.1, 0.1], dtype=np.float64)
b = np.array([0.1, 0.1], dtype=np.float64)
res = vec_ig(a, b)
```

The error that I receive is:

```
Compilation is falling back to object mode WITHOUT looplifting enabled because Function "vec_ig" failed type inference due to: No implementation of function Function(<built-in method gamma of numpy.random.mtrand.RandomState object at 0x7f7ae5b65840>) found for signature:
 
 >>> gamma(shape=float64, scale=float64)
 
There are 6 candidate implementations:
  - Of which 6 did not match due to:
  Overload in function 'gammavariate_impl': File: numba/cpython/randomimpl.py: Line 774.
    With argument(s): '(shape=float64, scale=float64)':
   Rejected as the implementation raised a specific error:
     TypingError: missing a required argument: 'alpha'
  raised from /home/devin/.cache/pypoetry/virtualenvs/pybuc-opbxzPBX-py3.9/lib64/python3.9/site-packages/numba/core/typing/templates.py:784
```

Expected behavior:

Return an array of size 2 with random draws from the gamma distribution.

### LINKED ISSUE #9073: Call by named arg broken for some `@overload` definitions
Numba has an issue with some internal `@overload` implementations whereby the definition of the implementation contains different argument names to the function that's being overload. In general this isn't noticed, but occasionally a user might try to call a function with named arguments for what would normally be the positional arguments. Example:

This overloads `np.geomspace()` it has a signature in NumPy like `np.geomspace(start, stop, num=50, <more arguments>)`. If if an `@overload` that exhibits the problem is created like:
```python
from numba import njit
import numpy as np
from numba.extending import overload

# NOTE: the name of the first arg is incorrect, NumPy API has `start`,
# this has `WRONG_NAME`.
@overload(np.geomspace)
def ol_geomspace(WRONG_NAME, stop, num=50):
    def impl(WRONG_NAME, stop, num=50):
        print("compiled and ran!")
    return impl

# ------------------------------------------------------------------------------
# This works fine
@njit
def foo1():
    np.geomspace(1, 2)

foo1()


# ------------------------------------------------------------------------------
# This works fine
@njit
def foo2():
    np.geomspace(1, stop=2)

foo2()


# ------------------------------------------------------------------------------
# This does not work, the overload expects `WRONG_NAME` but gets `start`.
# The user is just using the NumPy API as expected.
@njit
def foo3():
    np.geomspace(start=1, stop=2)

foo3()
```

Unfortunately there's quite a few functions defined as `@overload`s in Numba with this problem.

Two questions:
1. Where are these incorrectly defined functions in the current code base, they need fixing?
2. Should Numba have some mechanism to prevent incorrect definition for future `@overload` implementations?

As to 1. applying this patch locally seems to list most of the offending functions:
```diff
diff --git a/numba/core/extending.py b/numba/core/extending.py
--- a/numba/core/extending.py
+++ b/numba/core/extending.py
@@ -1,5 +1,6 @@
 import os
 import uuid
+import inspect
 import weakref
 import collections
 import functools
@@ -127,6 +128,33 @@ def overload(func, jit_options={}, strict=True, inline='never',
     # this requires sorting out the circular imports first.
 
     def decorate(overload_func):
+        try:
+            func_sig = inspect.signature(func)
+        except ValueError:
+            pass
+        else:
+            ol_sig = inspect.signature(overload_func)
+            x = list(func_sig.parameters.keys())
+            y = list(ol_sig.parameters.keys())
+            for a, b in zip(x[:len(y)], y):
+                if a != b:
+                    p = func_sig.parameters[a]
+                    if p.kind == p.POSITIONAL_ONLY:
+                        # probably a builtin written in C
+                        continue
+                    elif '*' in str(p): # probably *args or similar
+                        continue
+                    else:
+                        if not func.__module__.startswith("numba"):
+                            print(f"{func} from module '{getattr(func, '__module__')}' has mismatched sig.")
+                            print(f"    - expected: {func_sig}")
+                            print(f"    -      got: {ol_sig}")
+                            lineno = inspect.getsourcelines(overload_func)[1]
+                            tmpsrcfile = inspect.getfile(overload_func)
+                            srcfile = tmpsrcfile.replace(numba.__path__[0], '')
+                            print(f"from {srcfile}:{lineno}")
+                            print()
+                            break
         template = make_overload_template(func, overload_func, opts, strict,
                                           inline, prefer_literal, **kwargs)
         infer(template)

```

### PR DESCRIPTION
**Pull request by stuartarchibald**
This is a continuation of #9059, with thanks to @apmasell for this patch, authorship is preserved.

What this PR is about: Some of the Numba ``@overload``s for functions in NumPy and Python's built-ins were written using parameter names that did not match those used in API they were overloading. The result of this being that calling a function with such a mismatch using the parameter names as key-word arguments at the call site would result in a compilation error. This has now been universally fixed throughout the code base and a unit test is running with a best-effort attempt to prevent reintroduction of similar mistakes in the future.

Review order suggestion:
* Look at the combined diff of the original 5bcbb3a, cf75079 (adding test) and 587b15b (fixing further issues and de-duplicating code) to see that #9053 is addressed.
* Look solely at 86bf5cf next, this patch adds a function that scans all `@overload` API use and validates the function parameters. If you are happy with what this function does and the unit test that uses it, then the subsequent patches are just fixing the things it highlights as somewhat "mechanical" changes, each patch is isolated to a single module as per the commit message title.
* Look at 1bd2c6d which is a change log update.

Fixes #9053
Closes #9073

### CODE REVIEW

**Comment by stuartarchibald** [comment]
NOTE: The NumPy 1.21 tests are failing because:

The name of the first argument to `numpy.linalg.matrix_rank` changed between NumPy 1.21 (where it was `M`) and NumPy 1.22 (where it is `A`). This manifests as failing test (which was added in this PR, and is evidently working as it's catching this sort of thing):
```
numba.tests.test_extending.TestNumbaInternalOverloads.test_signatures_match_overloaded_api 
```
with exception:
```
ValueError: 
<function matrix_rank at 0x7f95fccbfa60> from module 'numpy.linalg' has mismatched sig.
    - expected: (M, tol=None, hermitian=False)
    -      got: (A, tol=None)
from /np/linalg.py:2495
```
As the base line NumPy version is being moved to 1.22 for Numba 0.58 (#9093), it's not worth fixing this, once that PR is merged the problem is resolved!

File: docs/upcoming_changes/9099.bug_fix.rst [hunk 1]
```diff
@@ -0,0 +1,69 @@
+Fix all ``@overload``s to use parameter names that match public APIs.
+=====================================================================
+
+Some of the Numba ``@overload``s for functions in NumPy and Python's built-ins
```
Comment (sklam) [inline docs/upcoming_changes/9099.bug_fix.rst:4]:
```suggestion
Some of the Numba ``@overload``\ s for functions in NumPy and Python's built-ins
```

**Review by sklam — COMMENTED** [review]
I've read through the changes and it looks okay.

Side-notes:  too bad this needs the `register_jittable` wrapper hack

File: docs/upcoming_changes/9099.bug_fix.rst [hunk 2]
```diff
@@ -0,0 +1,69 @@
+Fix all ``@overload``s to use parameter names that match public APIs.
```
Comment (sklam) [inline docs/upcoming_changes/9099.bug_fix.rst:1]:
```suggestion
Fix all ``@overload``\ s to use parameter names that match public APIs.
```

**Comment by sklam** [comment]
https://github.com/numba/numba/pull/9111 will fix the ReST syntax error at https://github.com/numba/numba/pull/9099#pullrequestreview-1555369076

**Comment by esc** [comment]
@sklam can this be closed, since it has been abandoned?

**Comment by sklam** [comment]
superseded by https://github.com/numba/numba/pull/9111

**Pull request closed without merging**